### DISCUSSION TITLE
Recommend way to type service class to use `this.emit()`?

### DISCUSSION THREAD

**Discussion by nmccrea** [feathersjs/feathers#3189] [Q&A]
In the documentation on [custom events](https://feathersjs.com/api/events.html#custom-events) it shows the following example code for TypeScript:

```typescript
class PaymentService {
  async create(data: any, params: Params) {
    const customer = await createStripeCustomer(params.user)
    this.emit('status', { status: 'created' })

    const payment = await createPayment(data)
    this.emit('status', { status: 'completed' })

    return payment
  }
}
```

But this will produce a TypeScript error at `this.emit`:

```
Property 'emit' does not exist on type 'PaymentService'.
```

I can solve this with `PaymentService extends EventEmitter`, but it seems like it would be better if there was a Feathers typing solution. I've been unsuccessful solving the issue using Feathers' types. For example, the `FeathersService` type includes the `emit()` addon but since it is only an interface, `PaymentService implements FeathersService<...>` doesn't actually endow the class with this property. Any recommendations on the best way to type this?

**Comment by daffl** [comment]
This is a good question, extending from EventEmitter seems like the cleanest solution to me. I found it quite difficult in TypeScript to deal with extended classes that have additional functionality.

**Reply by mvarendorff** [reply]
How would one do this in Dove when also extending `KnexService`? Is there any official solution to this or should one just roll custom types to slap on?

**Reply by mvarendorff** [reply]
I am currently using
```ts
declare module "@feathersjs/knex" {
  interface KnexService extends EventEmitter {}
}
```
as a workaround but I am open for alternative solutions!
